src/core/adt/show.js

```diff
@@ -0,0 +1,44 @@
+const { tagSymbol, typeSymbol } = require('./core');
+
+const objectToKeyValuePairs = (object) =>
+  Object.keys(object).map((key) => `${key}: ${showValue(object[key])}`).join(', ');
+
+const plainObjectToString = function() {
+  return `{ ${objectToKeyValuePairs(this)} }`;
+};
+
+const arrayToString = function() {
+  return `[${this.map(showValue).join(', ')}]`;
+};
+
+const nullToString = () => 'null';
+
+const objectToString = (object) => {
+  return object === null                                ?  nullToString
+  :      object.toString === Array.prototype.toString   ?  arrayToString
+  :      object.toString === Object.prototype.toString  ?  plainObjectToString
+  :      /* otherwise */                                   object.toString;
+};
+
+const showValue = (value) =>
+  typeof value !== 'object' ? JSON.stringify(value) : objectToString(value).call(value);
```

Comment:
We need to handle `undefined` here too, since JSON.stringify doesn't handle it :/